lib/shared/settings.js

```diff
@@ -0,0 +1,150 @@
+ * @function
+ * @private
+ *
+ * @returns {String} configuration file path
+ *
+ * @example
+ * const configurationFilePath = settings.getConfigurationFilePath();
+ */
+exports.getConfigurationFilePath = () => {
+  const hiddenFilePrefix = os.platform() === 'win32' ? '_' : '.';
+  return path.join(os.homedir(), `${hiddenFilePrefix}${packageJSON.name}rc`);
+};
+
+/**
+ * @summary Check if a setting value is invalid
+ * @function
+ * @private
+ *
+ * @description
+ * Our settings implementation only supports primitive values.
+ *
+ * @param {*} value - settings value
+ * @returns {Boolean} whether the setting value is invalid
+ *
+ * @example
+ * if (isInvalidSettingValue([ 1, 2, 3 ])) {
+ *   console.log('This setting value is invalid');
+ * }
+ */
+const isInvalidSettingValue = _.overSome([ _.isArray, _.isObject ]);
+
+/**
+ * @summary Read all settings from a configuration file
+ * @function
+ * @public
+ *
+ * @param {String} configurationFilePath - configuration file path
+ * @fulfil {Object} - settings object
+ * @returns {Promise}
+ *
+ * @example
+ * settings.readAll(settings.getConfigurationFilePath()).then((data) => {
+ *   console.log(data);
+ * });
+ */
+exports.readAll = (configurationFilePath) => {
+  return fs.readFileAsync(configurationFilePath, {
+    encoding: 'utf8'
+  }).then((contents) => {
+    if (_.isEmpty(_.trim(contents))) {
+      return {};
+    }
+
+    const result = _.attempt(JSON.parse, contents);
+
+    if (_.isError(result)) {
+      throw errors.createUserError({
+        title: `Invalid settings in ${configurationFilePath}`,
+        description: 'Please fix or delete this file to continue'
+      });
+    }
+
+    return _.omitBy(result, isInvalidSettingValue);
+  }).catch((error) => {
+    if (error.code === 'EPERM' || error.code === 'EACCES') {
+      throw errors.createUserError({
+        title: `Can't access settings in ${configurationFilePath}`,
+        description: 'Please fix or delete this file to continue'
+      });
+    }
+
+    if (error.code === 'ENOENT') {
+      return {};
+    }
+
+    throw error;
+  });
+};
+
+/**
+ * @summary Write all settings to a configuration file
+ * @function
+ * @public
+ *
+ * @description
+ * Note that this function will completely override any settings
+ * on the configuration file.
+ *
+ * @param {String} configurationFilePath - configuration file path
+ * @param {Object} settings - settings object
+ * @returns {Promise}
+ *
+ * @example
+ * settings.writeAll(settings.getConfigurationFilePath(), {
+ *   foo: 'bar'
+ * }).then(() => {
+ *   console.log('Done!');
+ * });
+ */
+exports.writeAll = (configurationFilePath, settings) => {
+  const invalidSettings = _.pickBy(settings, isInvalidSettingValue);
+  const JSON_INDENTATION_SPACES = 2;
+  const serialisedData = JSON.stringify(settings, null, JSON_INDENTATION_SPACES);
+
+  if (!_.isString(serialisedData) || !_.isPlainObject(settings) || !_.isEmpty(invalidSettings)) {
```

Comment:
Actually I'm wrong! I tried `_.isPlainObject(function(){})` again and it returns `false`—I must've prefixed it with `!` by mistake, copying the code. This means we *can* remove the first predicate in the or-chain because `JSON.stringify(() => {})` → `undefined`, `!_.isString(undefined)` → `false`, and `!_.isPlainObject(() => {})` → `false`, hence `false || false` → `false`.